src/main/java/io/airlift/slice/Slice.java

```diff
@@ -1194,8 +1193,7 @@ public String toString(int index, int length, Charset charset)
         if (base instanceof byte[]) {
             return new String((byte[]) base, (int) ((address - ARRAY_BYTE_BASE_OFFSET) + index), length, charset);
         }
-        // direct memory can only be converted to a string using a ByteBuffer
-        return decodeString(toByteBuffer(index, length), charset);
+        return new String(getBytes(index, length), charset);
```

Comment:
I'm pretty sure this will result in an extra copy.

Reply:
The existing code also copies because it's using a mutable, untrusted `CharBuffer`.